File: woodstorage/src/main/java/com/jordylangen/woodstorage/WoodStorageFactory.java
package com.jordylangen.woodstorage;

import android.content.Context;

import com.jordylangen.woodstorage.storage.StorageFactory;

import io.reactivex.processors.PublishProcessor;

public final class WoodStorageFactory {

    private static WoodStorageWorker WORKER;

    public static WoodStorageTree getInstance(Context context) {
        return getInstance(context, new StorageFactory());
    }

    public static WoodStorageTree getInstance(Context context, StorageFactory storageFactory) {
        PublishProcessor<LogEntry> publishSubject = PublishProcessor.create();
        WoodStorageTree tree = new WoodStorageTree(publishSubject);

        if (WORKER != null) {
            stop();
        }

        WORKER = new WoodStorageWorker(storageFactory.create(context), publishSubject);
        WORKER.start();

        return tree;
    }

    private static synchronized void stop() {
        WORKER.stop();
    }

    public static WoodStorageWorker getWorker() {
        return WORKER;
    }
}

File: woodstorage/src/main/java/com/jordylangen/woodstorage/WoodStorageFactory.kt
package com.jordylangen.woodstorage

import com.jordylangen.woodstorage.storage.LogEntry
import com.jordylangen.woodstorage.storage.StorageFactory

import io.reactivex.processors.PublishProcessor
import java.io.File

class WoodStorageFactory {
    companion object {
        private val publishSubject = PublishProcessor.create<LogEntry>()
        private val tree = WoodStorageTree(publishSubject)

        @JvmStatic
        var worker: WoodStorageWorker? = null
            private set

        @JvmStatic
        fun getInstance(directory: File, storageFactory: StorageFactory = StorageFactory()): WoodStorageTree {
            stop()
            worker = WoodStorageWorker(storageFactory.create(directory), publishSubject).also { it.start() }

            return tree
        }

        @Synchronized
        @JvmStatic
        private fun stop() = worker?.stop()
    }
}
